### PR TITLE
ir:feature - handling rhs identifiers

### PR DESCRIPTION
**Pull request by nathanmartinszup**
In some scenarios the value of rhs can be another identifier, this
scenario was not dealt previously in the IR. Now when there is a local
or global variable that uses another variable as a value, the original value of
the variable is used. For parameters we just set the value to nil.

Signed-off-by: Nathan Martins <nathan.martins@zup.com.br>

**- What I did**

**- How to verify it**

**- Description for the changelog**

### CODE REVIEW

File: internal/ir/builder.go [hunk 1]
```diff
@@ -233,12 +233,63 @@ func (b *builder) assign(fn *Function, lhs, rhs ast.Expr, syntax *ast.AssignStmt
...
+//
+// IR representation:
+//
+//	c = "y"
+//	d = "y"
+//	e = nil
+//	f = "x"
+//
+func (b *builder) identAssign(fn *Function, lhs, rhs *ast.Ident, syntax *ast.AssignStmt) {
+	if local, ok := fn.Locals[rhs.Name]; ok {
```
Comment (matheusalcantarazup) [inline internal/ir/builder.go:296]: Use `fn.lookup` instead to get a function local variable.

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]
I don't think that is good set a variable value to nil if the variable value points to a parameter, for example, the following code
```
function f(a) {
    const b = a
}
```

Should generate this IR:
```
func f(a):
0:                                                                         entry
	b = a
```

With this in mind, I think that the only change that should made is change the logic of `identAssign` function:
- Check if the identifier is a function local variable (It's already implemented)
- Check if the identifier is a parameter of function, if yes, them set the variable value to this parameter (Should be implemented)
- Check if the identifier is a global variable, if yes, them parse the global variable value to be the local variable value(it's already implemented)

So, basically we should iterate over the function parameter before and outside the global variable checker, something like this:

```go
for _, param := range fn.Signature.Params {
	if rhs.Name == param.name {
		fn.addNamedLocal(lhs.Name, param, syntax)

		return
	}
}

if global, ok := fn.File.Members[rhs.Name].(*Global); ok {
	fn.addNamedLocal(lhs.Name, b.expr(fn, global.Value, false /*expand*/), syntax)

	return
}
```

Btw, this PR seems pretty good, congrats.

File: internal/ir/builder.go (hunk 1 above)
Comment (nathanmartinszup) [reply internal/ir/builder.go]: Done!

File: internal/ir/builder.go [hunk 2]
```diff
@@ -257,12 +257,63 @@ func (b *builder) assign(fn *Function, lhs, rhs ast.Expr, syntax *ast.AssignStmt
 
 			return
 		}
+
+		if rhs, isIdent := rhs.(*ast.Ident); isIdent {
+			b.identAssign(fn, lhs, rhs, syntax)
+
+			return
+		}
```
Comment (matheusalcantarazup) [inline internal/ir/builder.go:265]: I think that this could be changed to a type switch case, what do you think?

File: internal/ir/builder.go [hunk 3]
```diff
@@ -257,12 +257,63 @@ func (b *builder) assign(fn *Function, lhs, rhs ast.Expr, syntax *ast.AssignStmt
...
+// source code:
+//
+// const a = "x"
+//
+// 	function f1(b) {
+//    	const c = "x"
+//    	const d = c
+//    	const e = b
+//    	const f = a
+//	}
```
Comment (matheusalcantarazup) [inline internal/ir/builder.go:286]: Fix the indentation here.

File: internal/ir/builder.go [hunk 4]
```diff
@@ -257,12 +257,63 @@ func (b *builder) assign(fn *Function, lhs, rhs ast.Expr, syntax *ast.AssignStmt
...
+//
+// IR representation:
+//
+//	c = "y"
+//	d = "y"
+//	e = nil
```
Comment (matheusalcantarazup) [inline internal/ir/builder.go:292]: Update the comment to `e = b`

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]

File: internal/ir/builder.go [hunk 5]
```diff
@@ -252,15 +252,69 @@ func (b *builder) assignStmt(fn *Function, lhss, rhss []ast.Expr, syntax *ast.As
...
-			fn.emit(b.funcLit(fn, lhs.Name, closure))
+		b.createValueByRHSType(fn, lhs, rhs, syntax)
+	default:
+		unsupportedNode(lhs)
+	}
+}
+
+// createValueByRHSType according to the type of the right operand, a different function will be executed to create
+// the var.
+func (b *builder) createValueByRHSType(fn *Function, lhs *ast.Ident, rhs ast.Expr, syntax *ast.AssignStmt) {
```
Comment (matheusalcantarazup) [inline internal/ir/builder.go:263]: I think that we can improve this function name and docs:

```go
// assingValue create a new Value according to the Value yielded from rhs expression.
//
// If the rhs expression is a *ast.FuncLit a new Closure Instruction will be created and emited
// to fn. If the rhs expression is a identifier, the identifier value will be resolve to set 
// the identifier value as the value of lhs identifier. Otherwise a new named local variable
// will be created using the Value created from rhs expression.
func (b *builder) assignValue(fn *Function, lhs *ast.Ident, rhs ast.Expr, syntax *ast.AssignStmt)
```

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]

File: internal/testdata/expected/javascript/ir/statements.js.out [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+# Location: statements.js:30:0
+func TryStatement():
+0:                                                                         entry
+
+# Name: WhileStatement
+# File: statements.js
+# Location: statements.js:42:0
+func WhileStatement():
+0:                                                                         entry
+
```
Comment (matheusalcantarazup) [inline internal/testdata/expected/javascript/ir/statements.js.out:63]: Remove these changes, we don't support these statements on IR. So tests should sill be failing.

**Review by matheusalcantarazup — CHANGES_REQUESTED** [review]
Sorry for these broken reviews, i think this is the last point

File: internal/testdata/expected/javascript/ir/statements.js.out (hunk 6 above)
Comment (nathanmartinszup) [reply internal/testdata/expected/javascript/ir/statements.js.out:63]: Wouldn't it be worth leaving it just to avoid confusion for other people working on the engine? I ended up wasting some time trying to understand why my changes broke the tests.

We can leave a TODO that this IR is not valid and must be updated. What do you think?

Comment (matheusalcantarazup) [reply internal/testdata/expected/javascript/ir/statements.js.out:63]: We are expecting that the IR tests fail because we don't support all nodes from ast package on IR. So since the same sample code is used to test the AST parsing and IR parsing, the IR tests should fail. The file `internal/semantic_test.go` contains a basic documentation about these tests.
